File: dqb-client/src/QueryExpression/Visitors/QueryExpressionVisitor.ts
import QueryExpression from "../QueryExpression";
import IQueryExpressionVisitor from "./IQueryExpressionVisitor";
import { format } from "sql-formatter";

export const DEFAULT_SQL_IDENTIFIERS: string[] = ["[", "]"];

export const DEFAULT_PREDICATE_OPERATORS: any = {
  Equals: "=",
  Like: "LIKE",
  In: "IN",
  GreaterThan: ">",
  GreaterThanOrEqualTo: ">=",
  LessThan: "<",
  LessThanOrEqualTo: "<=",
};

class QueryExpressionVisitor implements IQueryExpressionVisitor {
  _scope: any[] = [];
  _path: Pick<IQueryExpression, "_type">[] = [];

  _sqlBuilder: string[] = [];

  _logVisits: boolean = false;

  Reset = () => (this._sqlBuilder.length = 0);

  GetSql = () => {
    return this._sqlBuilder.join("");
  };

  Quotename = (name: string) => {
    return `${DEFAULT_SQL_IDENTIFIERS[0]}${name}${DEFAULT_SQL_IDENTIFIERS[1]}`;
  };

  Visit = (expression: IQueryExpression) => {
    this._scope.push(expression);
    this._path.push(expression._type as any);

    if (this._logVisits) console.log("Path:", this._path.join(" -> "));

    /// Call the correct `Visit_` function based on `expression._type`
    switch (expression._type) {
      case "table":
        this.Visit_TableQE(expression as ITableQE);
        break;
      case "select":
      case "single":
        this.Visit_SelectQE(expression as ISelectQE);
        break;
      case "column":
        this.Visit_ColumnQE(expression as IColumnQE);
        break;
      case "binary":
        this.Visit_BinaryQE(expression as IBinaryQE);
        break;
      case "constant":
        this.Visit_ConstantQE(expression as IConstantQE);
        break;
      case "join":
        this.Visit_JoinQE(expression as IJoinQE);
        break;
      case "predicate":
        this.Visit_PredicateQE(expression as IPredicateQE);
        break;
      case "and":
        this.Visit_AndQE(expression as IAndQE);
        break;
      case "or":
        this.Visit_OrQE(expression as IOrQE);
        break;
      case "cast":
        this.Visit_CastQE(expression as ICastQE);
        break;
      case "concat":
        this.Visit_ConcatQE(expression as IConcatQE);
        break;
      case "ternary":
        this.Visit_TernaryQE(expression as ITernaryQE);
        break;
      case "casecondition":
        this.Visit_CaseConditionQE(expression as ICaseConditionQE);
        break;
      case "case":
        this.Visit_CaseQE(expression as ICaseQE);
        break;
      case "groupby":
        this.Visit_GroupByQE(expression as IGroupByQE);
        break;
      case "aggregate":
        this.Visit_AggregateQE(expression as IAggregateQE);
        break;
      case "orderby":
        this.Visit_OrderByQE(expression as IOrderByQE);
        break;
      default:
        throw new Error(`Invalid expression type: ${expression._type}`);
    }
    this._scope.pop();
    this._path.pop();
  };

  Visit_TableQE(table: ITableQE) {
    if (this._logVisits) {
      console.log("Visit_TableQE", table);
    }
    if (table.catalog) this._sqlBuilder.push(this.Quotename(table.catalog));
    if (table.schema) this._sqlBuilder.push(this.Quotename(table.schema));
    this._sqlBuilder.push(this.Quotename(table.table));
    if (table.alias !== null)
      this._sqlBuilder.push(` AS ${this.Quotename(table.alias)}`);
  }

  Visit_SelectQE(select: ISelectQE) {
    if (this._logVisits) {
      console.log("Visit_SelectQE", select);
    }
    /// Add open paren for nested select
    if (this._scope.length > 1) this._sqlBuilder.push("(");
    this._sqlBuilder.push("SELECT\n");
    /// Add columns
    for (var i = 0; i < select.columns.length; i++) {
      new QueryExpression(select.columns[i]).Accept(this);

      if (i < select.columns.length - 1) this._sqlBuilder.push(",");
      this._sqlBuilder.push("\n");
    }

    this._sqlBuilder.push("FROM ");
    if (select.fromClause != null)
      new QueryExpression(select.fromClause).Accept(this);

    /// Add optional "JOIN" clause(s)
    if (select.joinClause != null && select.joinClause.length > 0) {
      select.joinClause.forEach((joinQE) => {
        if (joinQE._type === "join") new QueryExpression(joinQE).Accept(this);
      });
    }

    /// Add optional "WHERE" clause
    if (select.whereClause != null) {
      this._sqlBuilder.push(" WHERE ");
      new QueryExpression(select.whereClause).Accept(this);
    }

    /// Add optional "GROUB BY" clause
    if (select.groupByClause != null)
      new QueryExpression(select.groupByClause).Accept(this);

    /// Add optional "ORDER BY" clause(s)
    if (select.orderByClause != null && select.orderByClause.length > 0) {
      this._sqlBuilder.push(" ORDER BY ");
      for (var i = 0; i < select.orderByClause.length; i++) {
        new QueryExpression(select.orderByClause[i]).Accept(this);

        if (i < select.orderByClause.length - 1) this._sqlBuilder.push(", ");
      }
    }

    /// Add close paren and optional alias for nested select
    if (this._scope.length > 1) {
      this._sqlBuilder.push(")");
      if (select.alias)
        this._sqlBuilder.push(` AS ${this.Quotename(select.alias)}`);
    }
  }

  Visit_ColumnQE(column: IColumnQE) {
    if (this._logVisits) {
      console.log("Visit_ColumnQE");
    }
    if (
      this._path.length < 2 ||
      this._path[this._path.length - 2] === ("select" as any)
    )
      this._sqlBuilder.push("\t");
    if (column.table !== null)
      this._sqlBuilder.push(`${this.Quotename(column.table)}.`);
    this._sqlBuilder.push(this.Quotename(column.name));
    this.AddColumnAlias(column);
  }

  Visit_BinaryQE(binary: IBinaryQE) {
    if (this._logVisits) {
      console.log("Visit_BinaryQE");
    }
    if (
      this._path.length < 2 ||
      this._path[this._path.length - 2] === ("select" as any)
    )
      this._sqlBuilder.push("\t");
    this._sqlBuilder.push("(");
    new QueryExpression(binary.left).Accept(this);
    this._sqlBuilder.push(` ${binary.operator} `);
    new QueryExpression(binary.right).Accept(this);
    this._sqlBuilder.push(")");
    this.AddColumnAlias(binary);
  }

  Visit_ConstantQE(constant: IConstantQE) {
    if (this._logVisits) {
      console.log("Visit_ConstantQE");
    }
  }

  Visit_JoinQE(join: IJoinQE) {
    if (this._logVisits) {
      console.log("Visit_JoinQE");
    }
    this._sqlBuilder.push(` ${this.GetJoinTypeSymbol(join.joinType)} `);
    new QueryExpression(join.target).Accept(this);
    this._sqlBuilder.push(" ON ");
    new QueryExpression(join.condition).Accept(this);
  }

  Visit_PredicateQE(predicate: IPredicateQE) {
    if (this._logVisits) {
      console.log("Visit_PredicateQE", predicate);
    }
    new QueryExpression(predicate.left).Accept(this);
    this._sqlBuilder.push(
      ` ${this.GetPredicateOperatorSymbol(predicate.operator)} `
    );
    new QueryExpression(predicate.right).Accept(this);
  }

  Visit_AndQE(and: IAndQE) {
    if (this._logVisits) {
      console.log("Visit_AndQE");
    }
    this._sqlBuilder.push("(");
    for (var i = 0; i < and.expressions.length; i++) {
      new QueryExpression(and.expressions[i]).Accept(this);
      if (i < and.expressions.length - 1) {
        this._sqlBuilder.push(" AND ");
      }
    }
    this._sqlBuilder.push(")");
  }

  Visit_OrQE(or: IOrQE) {
    if (this._logVisits) {
      console.log("Visit_OrQE");
    }
    this._sqlBuilder.push("(");
    for (var i = 0; i < or.expressions.length; i++) {
      new QueryExpression(or.expressions[i]).Accept(this);
      if (i < or.expressions.length - 1) {
        this._sqlBuilder.push(" OR ");
      }
    }
    this._sqlBuilder.push(")");
  }

  Visit_CastQE(cast: ICastQE) {
    if (this._logVisits) {
      console.log("Visit_CastQE");
    }
    this._sqlBuilder.push("CAST(");
    new QueryExpression(cast.operand).Accept(this);
    this._sqlBuilder.push(` AS ${cast.targetType}`);
    this.AddColumnAlias(cast);
  }

  Visit_ConcatQE(concat: IConcatQE) {
    if (this._logVisits) {
      console.log("Visit_ConcatCE");
    }
    this._sqlBuilder.push("CONCAT(");
    for (var i = 0; i < concat.arguments.length; i++) {
      new QueryExpression(concat.arguments[i]).Accept(this);
      if (i < concat.arguments.length - 1) this._sqlBuilder.push(", ");
    }
    this._sqlBuilder.push(")");
    this.AddColumnAlias(concat);
  }

  Visit_TernaryQE(ternary: ITernaryQE) {
    if (this._logVisits) {
      console.log("Visit_TernaryQE");
    }
    this._sqlBuilder.push("CASE WHEN ");
    new QueryExpression(ternary.predicate).Accept(this);
    this._sqlBuilder.push(" THEN ");
    new QueryExpression(ternary.trueExpression).Accept(this);
    this._sqlBuilder.push(" ELSE ");
    new QueryExpression(ternary.falseExpression).Accept(this);
    this._sqlBuilder.push(" END");
    this.AddColumnAlias(ternary);
  }

  Visit_CaseConditionQE(caseCondition: ICaseConditionQE) {
    if (this._logVisits) {
      console.log("Visit_CaseConditionQE");
    }
    this._sqlBuilder.push("WHEN ");
    new QueryExpression(caseCondition.predicate).Accept(this);
    this._sqlBuilder.push(" THEN ");
    new QueryExpression(caseCondition.then).Accept(this);
  }

  Visit_CaseQE(caseQE: ICaseQE) {
    if (this._logVisits) {
      console.log("Visit_CaseConditionQE");
    }
    this._sqlBuilder.push("CASE ");
    caseQE.conditions.forEach((caseCondition: ICaseConditionQE) => {
      new QueryExpression(caseCondition).Accept(this);
    });

    this._sqlBuilder.push(" ELSE ");
    new QueryExpression(caseQE.default).Accept(this);
    this._sqlBuilder.push(" END");
    this.AddColumnAlias(caseQE);
  }

  Visit_GroupByQE(groupByQE: IGroupByQE) {
    if (this._logVisits) {
      console.log("Visit_CaseConditionQE");
    }
  }

  Visit_AggregateQE(aggregate: IAggregateQE) {
    if (this._logVisits) {
      console.log("Visit_CaseConditionQE");
    }
    this._sqlBuilder.push(`${aggregate.function.toString()}(`);
    new QueryExpression(aggregate.expression).Accept(this);
    this._sqlBuilder.push(")");
    this.AddColumnAlias(aggregate);
  }

  Visit_OrderByQE(orderBy: IOrderByQE) {
    if (this._logVisits) {
      console.log("Visit_CaseConditionQE");
    }
    new QueryExpression(orderBy.column).Accept(this);
    this._sqlBuilder.push(` ${orderBy.direction}`);
  }

  AddColumnAlias(columnExpression: IQueryExpression) {
    if (
      this._path.length < 2 ||
      this._path[this._path.length - 2] === ("select" as any)
    ) {
      var alias = this.GetColumnAlias(columnExpression);
      if (alias !== null && alias !== undefined)
        this._sqlBuilder.push(` AS ${this.Quotename(alias)}`);
    }
  }

  GetColumnAlias(columnExpression: any) {
    if (columnExpression.alias !== null) return columnExpression.alias;
    else if (columnExpression._type === "column") return columnExpression.name;
    return null;
  }

  GetJoinTypeSymbol(joinType: string) {
    return joinType.toUpperCase() + " JOIN";
  }

  GetPredicateOperatorSymbol(predicateOperator: string) {
    if (Object.keys(DEFAULT_PREDICATE_OPERATORS).includes(predicateOperator))
      return DEFAULT_PREDICATE_OPERATORS[predicateOperator];
    return "=";
  }
}

export default QueryExpressionVisitor;

File: DotnetQueryBuilder.Core/Models/QueryExpressionVisitors/QueryExpressionVisitor.cs
using System.Text;
using BasicSQLFormatter;

namespace DotnetQueryBuilder.Core;
public class QueryExpressionVisitor : IQueryExpressionVisitor
{
    public readonly List<QueryExpression> _scope = new List<QueryExpression>();
    public readonly StringBuilder _sqlBuilder = new StringBuilder();
    public readonly List<(string target, string? message)> _errors = new List<(string target, string? message)>();
    public readonly List<string> _tableExpressionAliases = new List<string>();
    public virtual string _sqlIdentifiers { get => "[]"; }
    private readonly List<dynamic> _constants = new List<dynamic>();
    public void Reset()
    {
        _sqlBuilder.Clear();
    }
    public string GetSql()
    {
        return new SQLFormatter(_sqlBuilder.ToString()).Format();
    }
    public void OpenScope(QueryExpression current)
    {
        _scope.Add(current);
    }
    public void CloseScope(QueryExpression current)
    {
        _scope.RemoveAt(_scope.LastIndexOf(current));
    }
    public List<Type> GetPath() => _scope.Select(i => i.GetType()).ToList();
    public List<Type> _path { get => GetPath(); }
    public string GetPathString() => string.Join(',', _scope.Select(i => i.GetType().Name));
    private void AddColumnAlias(IColumnExpression columnExpression)
    {
        if (_path.Count < 2 || _path[_path.Count - 2] == typeof(SelectQE))
        {
            var alias = GetColumnAlias(columnExpression);
            if (!string.IsNullOrEmpty(alias))
                _sqlBuilder.Append($" AS {_sqlIdentifiers[0]}{alias}{_sqlIdentifiers[1]}");
        }
    }
    private string? GetColumnAlias(IColumnExpression columnExpression)
    {
        if (!string.IsNullOrEmpty(columnExpression.Alias))
            return columnExpression.Alias;
        else if (columnExpression.GetType() == typeof(ColumnQE))
            return ((ColumnQE)(columnExpression)).Name;
        return null;
    }
    public virtual void Visit(TableQE table)
    {
        if (!string.IsNullOrEmpty(table.Catalog))
            _sqlBuilder.Append($"{_sqlIdentifiers[0]}{table.Catalog}{_sqlIdentifiers[1]}.");
        if (!string.IsNullOrEmpty(table.Schema))
            _sqlBuilder.Append($"{_sqlIdentifiers[0]}{table.Schema}{_sqlIdentifiers[1]}.");
        _sqlBuilder.Append($"{_sqlIdentifiers[0]}{table.Table}{_sqlIdentifiers[1]}");
        var alias = string.IsNullOrEmpty(table.Alias) ? $"t{_tableExpressionAliases.Count}" : table.Alias;
        _sqlBuilder.Append($" AS {_sqlIdentifiers[0]}{alias}{_sqlIdentifiers[1]}");
    }
    public virtual void Visit(ITableExpression tableExpression)
    {
        Console.WriteLine("Visit ITableExpression");
    }
    public virtual void Visit(SelectQE select)
    {
        /// Add open paren for nested select
        if (_scope.Count > 1)
            _sqlBuilder.Append("(");
        _sqlBuilder.Append("SELECT ");

        /// Add columns
        for (int i = 0; i < select.Columns.Count(); i++)
        {
            select.Columns.ToArray()[i].Accept(this);

            if (i < select.Columns.Count() - 1)
                _sqlBuilder.Append(", ");
        }

        _sqlBuilder.Append(" FROM ");
        if (select.FromClause != null)
            select.FromClause.Accept(this);

        /// Add optional "JOIN" clause(s)
        if (select.JoinClause != null && select.JoinClause.Count() > 0)
        {
            foreach (var joinQE in select.JoinClause)
            {
                if (joinQE.GetType() == typeof(JoinQE))
                    joinQE.Accept(this);
                else
                    _errors.Add((select.ExpressionType, $"Query Expression {joinQE.GetType().Name} is not valid for `SelectQE.Join`"));
            }
        }

        /// Add optional "WHERE" clause
        if (select.WhereClause != null)
        {
            _sqlBuilder.Append(" WHERE ");
            select.WhereClause.Accept(this);
        }

        /// Add optional "GROUB BY" clause
        if (select.GroupByClause != null)
            select.GroupByClause.Accept(this);

        /// Add optional "ORDER BY" clause(s)
        if (select.OrderByClause != null && select.OrderByClause.Count() > 0)
        {
            _sqlBuilder.Append(" ORDER BY ");
            for (int i = 0; i < select.OrderByClause.Count(); i++)
            {
                select.OrderByClause.ToArray()[i].Accept(this);

                if (i < select.OrderByClause.Count() - 1)
                    _sqlBuilder.Append(", ");
            }
        }

        /// Add close paren and optional alias for nested select
        if (_scope.Count > 1)
        {
            _sqlBuilder.Append(")");
            if (!string.IsNullOrEmpty(select.Alias))
                _sqlBuilder.Append($" AS {_sqlIdentifiers[0]}{select.Alias}{_sqlIdentifiers[1]}");
        }


    }
    public virtual void Visit(ColumnQE column)
    {
        if (!string.IsNullOrEmpty(column.Table))
            _sqlBuilder.Append($"{_sqlIdentifiers[0]}{column.Table}{_sqlIdentifiers[1]}.");
        _sqlBuilder.Append($"{_sqlIdentifiers[0]}{column.Name}{_sqlIdentifiers[1]}");
        AddColumnAlias(column);
    }
    public virtual void Visit(OrderByQE orderBy)
    {
        orderBy.Column.Accept(this);
        _sqlBuilder.Append($" {orderBy.Direction}");
    }
    public virtual void Visit(AggregateQE aggregate)
    {
        _sqlBuilder.Append($"{aggregate.Function.ToString()}(");
        aggregate.Expression.Accept(this);
        _sqlBuilder.Append(")");
        AddColumnAlias(aggregate);
    }
    public virtual void Visit(BinaryQE binary)
    {
        _sqlBuilder.Append("(");
        binary.Left.Accept(this);
        _sqlBuilder.Append($" {GetBinaryOperatorSymbol(binary.Operator)} ");
        binary.Right.Accept(this);
        _sqlBuilder.Append(")");
        AddColumnAlias(binary);
    }
    public virtual void Visit(ConstantQE constant)
    {
        // _constants.Add(constant.Value);
        // _sqlBuilder.Append($"@Const{(_constants.Count() - 1)}");
        if (constant.Value is int || constant.Value is short || constant.Value is long || constant.Value is decimal || constant.Value is float)
            _sqlBuilder.Append(constant.Value.ToString());
        else
            _sqlBuilder.Append($"'{constant.Value.ToString()}'");

        AddColumnAlias(constant);
    }
    public virtual void Visit(JoinQE join)
    {
        _sqlBuilder.Append($" {GetJoinTypeSymbol(join.JoinType)} ");
        join.Target.Accept(this);
        _sqlBuilder.Append(" ON ");
        join.Condition.Accept(this);
    }
    public virtual void Visit(GroupByQE groupBy)
    {
        if (groupBy.Columns.Count() > 0)
        {
            _sqlBuilder.Append(" GROUP BY ");
            for (int i = 0; i < groupBy.Columns.Count(); i++)
            {
                groupBy.Columns.ToArray()[i].Accept(this);
                if (i < groupBy.Columns.Count() - 1)
                    _sqlBuilder.Append(", ");
            }
        }
    }
    public virtual void Visit(PredicateQE predicate)
    {
        predicate.Left.Accept(this);
        _sqlBuilder.Append($" {GetPredicateOperatorSymbol(predicate.Operator)} ");
        predicate.Right.Accept(this);
    }
    public virtual void Visit(AndQE and)
    {
        _sqlBuilder.Append("(");
        for (int i = 0; i < and.Expressions.Count(); i++)
        {
            and.Expressions.ToArray()[i].Accept(this);
            if (i < and.Expressions.Count() - 1)
            {
                _sqlBuilder.Append(" AND ");
            }
        }
        _sqlBuilder.Append(")");
    }
    public virtual void Visit(OrQE or)
    {
        _sqlBuilder.Append("(");
        for (int i = 0; i < or.Expressions.Count(); i++)
        {
            or.Expressions.ToArray()[i].Accept(this);
            if (i < or.Expressions.Count() - 1)
            {
                _sqlBuilder.Append(" OR ");
            }
        }
        _sqlBuilder.Append(")");
    }
    public virtual void Visit(CastQE cast)
    {
        _sqlBuilder.Append("CAST(");
        cast.Operand.Accept(this);
        _sqlBuilder.Append($" AS {cast.TargetType}");
        AddColumnAlias(cast);
    }
    public virtual void Visit(ConcatQE concat)
    {
        _sqlBuilder.Append("CONCAT(");
        for (int i = 0; i < concat.Arguments.Count(); i++)
        {
            concat.Arguments.ToArray()[i].Accept(this);
            if (i < concat.Arguments.Count() - 1)
                _sqlBuilder.Append(", ");
        }
        _sqlBuilder.Append(")");
        AddColumnAlias(concat);
    }
    public virtual void Visit(TernaryQE ternary)
    {
        _sqlBuilder.Append("CASE WHEN ");
        ternary.Predicate.Accept(this);
        _sqlBuilder.Append(" THEN ");
        ternary.TrueExpression.Accept(this);
        _sqlBuilder.Append(" ELSE ");
        ternary.FalseExpression.Accept(this);
        _sqlBuilder.Append(" END");
        AddColumnAlias(ternary);
    }
    public virtual void Visit(CaseConditionQE caseCondition)
    {
        _sqlBuilder.Append("WHEN ");
        caseCondition.Predicate.Accept(this);
        _sqlBuilder.Append(" THEN ");
        caseCondition.Then.Accept(this);
    }
    public virtual void Visit(CaseQE caseQE)
    {
        _sqlBuilder.Append("CASE ");
        foreach (var caseCondition in caseQE.Conditions)
            caseCondition.Accept(this);
        _sqlBuilder.Append(" ELSE ");
        caseQE.Default.Accept(this);
        _sqlBuilder.Append(" END");
        AddColumnAlias(caseQE);
    }

    public static readonly Dictionary<BinaryQEOperator, string> BinaryOperatorSymbols =
        new Dictionary<BinaryQEOperator, string>() {
            {BinaryQEOperator.Add, "+"},
            {BinaryQEOperator.Subtract, "-"},
            {BinaryQEOperator.Muliply, "*"},
            {BinaryQEOperator.Divide, "/"}
        };
    public static readonly Dictionary<PredicateQEOperator, string> PredicateOperatorSymbols =
        new Dictionary<PredicateQEOperator, string>() {
            {PredicateQEOperator.Equals, "="},
            {PredicateQEOperator.GreaterThan, ">"},
            {PredicateQEOperator.GreaterThanOrEqualTo, ">="},
            {PredicateQEOperator.LessThan, "<"},
            {PredicateQEOperator.LessThanOrEqualTo, "<="},
            {PredicateQEOperator.Like, "LIKE"},
            {PredicateQEOperator.In, "IN"},
        };

    public static readonly Dictionary<JoinQEType, string> JoinTypeSymbols = new Dictionary<JoinQEType, string>() {
        {JoinQEType.INNER, "INNER JOIN"},
        {JoinQEType.LEFT, "LEFT JOIN"}
    };

    private static string GetPredicateOperatorSymbol(PredicateQEOperator op)
    {
        return PredicateOperatorSymbols[op];
    }

    private static string GetBinaryOperatorSymbol(BinaryQEOperator op)
    {
        return BinaryOperatorSymbols[op];
    }

    public static string GetJoinTypeSymbol(JoinQEType op)
    {
        return JoinTypeSymbols[op];
    }



}